x/src/lint/engine/mod.rs

```diff
@@ -0,0 +1,207 @@
+pub trait Linter: Send + Sync + fmt::Debug {
+    /// Returns the name of the linter.
+    fn name(&self) -> &'static str;
+}
+
+/// Represents common functionality among various `Context` instances.
+pub(super) trait LintContext<'l> {
+    /// Returns the kind of this lint context.
+    fn kind(&self) -> LintKind<'l>;
+
+    /// Returns a `LintSource` for this lint context.
+    fn source(&self, name: &'static str) -> LintSource<'l> {
+        LintSource::new(name, self.kind())
+    }
+}
+
+/// A lint formatter.
+///
+/// Lints write `LintMessage` instances to this.
+pub struct LintFormatter<'l, 'a> {
+    source: LintSource<'l>,
+    messages: &'a mut Vec<(LintSource<'l>, LintMessage)>,
+}
+
+impl<'l, 'a> LintFormatter<'l, 'a> {
+    pub fn new(
+        source: LintSource<'l>,
+        messages: &'a mut Vec<(LintSource<'l>, LintMessage)>,
+    ) -> Self {
+        Self { source, messages }
+    }
+
+    /// Writes a new lint message to this formatter.
+    pub fn write(&mut self, message: LintMessage) {
+        self.messages.push((self.source, message));
+    }
+}
+
+/// The run status of a lint.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum RunStatus<'l> {
+    /// This lint run was successful, with messages possibly written into the `LintFormatter`.
+    Executed,
+    /// This lint was skipped.
+    Skipped(SkipReason<'l>),
+}
+
+/// The reason for why this lint was skipped.
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum SkipReason<'l> {
+    /// This file was not valid UTF-8.
+    NonUtf8,
+    /// This extension was unsupported.
+    UnsupportedExtension(Option<&'l str>),
+    // TODO: Add more reasons.
+}
+
+/// A message raised by a lint.
+#[derive(Debug)]
+pub struct LintMessage {
+    level: LintLevel,
+    message: Cow<'static, str>,
+}
+
+impl LintMessage {
+    pub fn new(level: LintLevel, message: impl Into<Cow<'static, str>>) -> Self {
+        Self {
+            level,
+            message: message.into(),
+        }
+    }
+
+    pub fn level(&self) -> LintLevel {
+        self.level
+    }
+
+    pub fn message(&self) -> &str {
+        &self.message
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[allow(dead_code)]
+#[non_exhaustive]
+pub enum LintLevel {
+    Error,
+    Warning,
+    // TODO: add more levels?
+}
+
+impl fmt::Display for LintLevel {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            LintLevel::Error => write!(f, "ERROR"),
+            LintLevel::Warning => write!(f, "WARNING"),
+        }
+    }
+}
+
+/// Message source for lints.
+#[derive(Copy, Clone, Debug)]
+pub struct LintSource<'l> {
+    name: &'static str,
+    kind: LintKind<'l>,
+}
+
+impl<'l> LintSource<'l> {
+    pub(super) fn new(name: &'static str, kind: LintKind<'l>) -> Self {
+        Self { name, kind }
+    }
+
+    pub fn name(&self) -> &'static str {
+        self.name
+    }
+
+    pub fn kind(&self) -> LintKind<'l> {
+        self.kind
```

Comment:
I was tripped up here for a second till I realized LintKind was Copy :)

Reply:
Yeah, any suggestions for better name?

Reply:
This wasn't really an issue on the name, I just needed to realize that the type was copy and that this was ok.